### PR TITLE
stm32l1/i2c: initial update for the i2c peripheral driver

### PR DESCRIPTION
**Pull request by ReneHerthel**
- removes switch-cases
- added Config-array with information of the I2C devices
- faster access to the device information
- supports the stm32l1-CPU with the limifrog-v1 and the nucleo-l1
- supports now I2C_1 (not only I2C_0)

This PR is needed for testing the limifrog-v1 drivers.

### CODE REVIEW

**Comment by ReneHerthel** [comment]
@PeterKietzmann Here is the PR about the i2c driver we talkt about offline.

File: boards/limifrog-v1/include/periph_conf.h [hunk 1]
```diff
@@ -139,37 +139,27 @@ static const timer_conf_t timer_config[] = {
 #define I2C_IRQ_PRIO        1
 #define I2C_APBCLK          (36000000U)
 
+static const i2c_conf_t i2c_config[] = {
+    /* device, port, scl-, sda-pin-number, scl-, sda-AF, ER-IRQn, EV-IRQn */
+    {I2C1, GPIOB,  8,  9, 4, 4, I2C1_ER_IRQn, I2C1_EV_IRQn},
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: Please initialize all pins with the [pin generator macro](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32l1/include/periph_cpu.h#L46) to be independent of the port.

File: boards/limifrog-v1/include/periph_conf.h [hunk 2]
```diff
@@ -139,37 +139,27 @@ static const timer_conf_t timer_config[] = {
...
+    {I2C2, GPIOB, 10, 11, 4, 4, I2C2_ER_IRQn, I2C2_EV_IRQn},
+};
+
```
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: @ReneHerthel do you think there is a reasonable way to do the clock initialization in line 144/145 inside the driver? I mean, in the driver you now the I2C device and the port and the pin number and so on...

Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: Do you know what `I2C_0_EVT_ISR` in line 151 is good for?

File: boards/nucleo-l1/include/periph_conf.h [hunk 3]
```diff
@@ -238,26 +238,33 @@ static const timer_conf_t timer_config[] = {
  * @name I2C configuration
   * @{
  */
-#define I2C_NUMOF           (1U)
+#define I2C_NUMOF           (2U)
```
Comment (PeterKietzmann) [inline boards/nucleo-l1/include/periph_conf.h]: The same comments as above apply to this periph_conf

File: boards/limifrog-v1/include/periph_conf.h (hunk 1 above)
Comment (PeterKietzmann) [inline boards/limifrog-v1/include/periph_conf.h]: Ah and nearly forgot: there are also mux types defined [there](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32l1/include/periph_cpu.h#L70). Please use them

File: cpu/stm32l1/include/periph_cpu.h [hunk 4]
```diff
@@ -100,6 +100,20 @@ typedef struct {
...
+typedef struct {
+    I2C_TypeDef *dev;       /**< i2c device */
+    GPIO_TypeDef *port;     /**< i2c port */
+    uint8_t scl;            /**< scl pin number */
+    uint8_t sda;            /**< sda pin number */
+    uint8_t scl_af;         /**< scl alternate function value */
+    uint8_t sda_af;         /**< sda alternate function value */
+    uint8_t er_irqn;        /**< error IRQ */
+    uint8_t ev_irqn;        /**< event IRQ */
+} i2c_conf_t;
```
Comment (PeterKietzmann) [inline cpu/stm32l1/include/periph_cpu.h]: Please adapt the proposed changes in the periph_conf

File: cpu/stm32l1/periph/i2c.c [hunk 5]
```diff
@@ -34,17 +34,14 @@
 #define ENABLE_DEBUG    (0)
 #include "debug.h"
 
-/* guard file in case no I2C device is defined */
-#if I2C_NUMOF
-
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: Why did you remove that guard?

File: cpu/stm32l1/periph/i2c.c [hunk 6]
```diff
@@ -85,27 +81,31 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
-            I2C_0_CLKEN();
             I2C_0_PORT_CLKEN();
-            NVIC_SetPriority(I2C_0_ERR_IRQ, I2C_IRQ_PRIO);
-            NVIC_EnableIRQ(I2C_0_ERR_IRQ);
             break;
-#endif
+
+        case I2C_1:
+            I2C_1_PORT_CLKEN();
+            break;
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: As I proposed above: If you removed the need for these macros here, you could remove the whole switch case here. One remark in addition: Could you check if this is done by the GPIO configuration anyway?

File: cpu/stm32l1/periph/i2c.c [hunk 7]
```diff
@@ -85,27 +81,31 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
+
+    /* set IRQn priority */
+    NVIC_SetPriority(i2c_config[dev].er_irqn, I2C_IRQ_PRIO);
+
+    /* enable IRQn */
+    NVIC_EnableIRQ(i2c_config[dev].er_irqn);
+
     /* configure pins */
-    _pin_config(port, pin_scl, pin_sda);
+    _pin_config(dev);
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: @ReneHerthel please look at the GPIO API. I'd like to see it being used instead of reprogramming IO- and AF-initialization. Can you check that please? In that case you can exchange the whole ` _pin_config` function by some GPIO API calls

File: cpu/stm32l1/periph/i2c.c [hunk 8]
```diff
@@ -116,9 +116,9 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
         /* disable peripheral */
         i2c->CR1 &= ~I2C_CR1_PE;
         /* toggle both pins to reset analog filter */
-        _toggle_pins(port, pin_scl, pin_sda);
+        _toggle_pins(dev);
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: @ReneHerthel in line 113 an errata sheet was mentioned. I **assume** the whole block was copied from one CPU once and dragged along even if the errata doesn't apply to all stm32 CPUs. Can you please find out if the errata sheet applies to this CPU and if not, remove the whole `if (i2c->SR2 & I2C_SR2_BUSY)` block and test if driver still works?

File: cpu/stm32l1/periph/i2c.c [hunk 9]
```diff
@@ -142,91 +142,84 @@ static void _i2c_init(I2C_TypeDef *i2c, int ccr)
     i2c->CR1 |= I2C_CR1_PE;
 }
 
-static void _pin_config(GPIO_TypeDef *port, int pin_scl, int pin_sda)
+static void _pin_config(i2c_t dev)
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: Necessity of this function discussed above

File: cpu/stm32l1/periph/i2c.c [hunk 10]
```diff
@@ -142,91 +142,84 @@ static void _i2c_init(I2C_TypeDef *i2c, int ccr)
...
+
+        /* Set alternate function of the pin to I2C */
+        hl = (pin[i] < 8) ? 0 : 1;
+        port->AFR[hl] &= ~(0xf << ((pin[i] - (hl * 8)) * 4));
+        port->AFR[hl] |= (af[i] << ((pin[i] - (hl * 8)) * 4));
     }
 }
 
-static void _toggle_pins(GPIO_TypeDef *port, int pin_scl, int pin_sda)
+static void _toggle_pins(i2c_t dev)
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: Necessity of this function discussed above. Can be removed if the above mentioned errata sheet doesn't affect this cpu

File: cpu/stm32l1/periph/i2c.c [hunk 11]
```diff
@@ -142,91 +142,84 @@ static void _i2c_init(I2C_TypeDef *i2c, int ccr)
...
     /* set SDA to high */
-    port->ODR |= (1 << pin_sda);
+    port->ODR |= (1 << pin[1]);
 }
 
 int i2c_acquire(i2c_t dev)
 {
-    if (dev >= I2C_NUMOF) {
-        return -1;
-    }
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: Why did you remove the check?

File: cpu/stm32l1/periph/i2c.c [hunk 12]
```diff
@@ -467,99 +411,94 @@ int i2c_write_regs(i2c_t dev, uint8_t address, uint8_t reg, char *data, int leng
...
         case I2C_0:
             I2C_0_CLKEN();
             break;
-#endif
-#if I2C_1_EN
+
         case I2C_1:
             I2C_1_CLKEN();
             break;
-#endif
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: If there was a way to automatically enable/disable the clock for the concerned device, one would save the whole switch case.

File: cpu/stm32l1/periph/i2c.c [hunk 13]
```diff
@@ -467,99 +411,94 @@ int i2c_write_regs(i2c_t dev, uint8_t address, uint8_t reg, char *data, int leng
...
-#if I2C_1_EN
+
         case I2C_1:
-            while (I2C_0_DEV->SR2 & I2C_SR2_BUSY);
-            I2C_0_CLKDIS();
+            while (i2c_config[dev].dev->SR2 & I2C_SR2_BUSY);
+            I2C_1_CLKDIS();
             break;
-#endif
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: If there was a way to automatically enable/disable the clock for the concerned device, one would save the whole switch case.

**Comment by PeterKietzmann** [comment]
@ReneHerthel thanks for your work. Please try to address my comments and/or answer my questions. I know you just try to improve the driver but now that you are at it, we can try to optimize a bit more ;-) !

File: boards/limifrog-v1/include/periph_conf.h (hunk 2 above)
Comment (ReneHerthel) [inline boards/limifrog-v1/include/periph_conf.h]: @PeterKietzmann Well, now that you mention it, I think it's not used anywhere? It is defined in some other periph_conf files, but not used? hmm.. or did I missed something?

File: boards/limifrog-v1/include/periph_conf.h (hunk 1 above)
Comment (ReneHerthel) [inline boards/limifrog-v1/include/periph_conf.h]: @PeterKietzmann Done. But how to extract the pin and port number in the i2c driver? using the gpio api?

File: boards/limifrog-v1/include/periph_conf.h (hunk 2 above)
Comment (lebrush) [inline boards/limifrog-v1/include/periph_conf.h]: AFAIK the i2c on the stm32 can work either interrupt based or with busy waits, which is how it's currently implemented. The event interrupt handling is not used by any of the stm32 in RIOT :+1:

File: cpu/stm32l1/periph/i2c.c (hunk 5 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann I though the drivers are used now with "best knowledge", that the person who use them, knows how his board works? Or is it different with I2C ?

File: cpu/stm32l1/periph/i2c.c [hunk 14]
```diff
@@ -66,13 +63,12 @@ static mutex_t locks[] =  {
 
 int i2c_init_master(i2c_t dev, i2c_speed_t speed)
 {
-    I2C_TypeDef *i2c;
-    GPIO_TypeDef *port;
-    int pin_scl = 0, pin_sda = 0;
+    I2C_TypeDef *i2c = i2c_config[dev].dev;
```
Comment (lebrush) [inline cpu/stm32l1/periph/i2c.c]: This means that if `dev` is larger than the defined elements we are going to access someone else's other memory and going to be fun to debug... I think the old behaviour (verification) makes more sense for a general purpose driver

File: cpu/stm32l1/periph/i2c.c (hunk 7 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: Good point, I'll check that.

File: cpu/stm32l1/periph/i2c.c (hunk 11 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann As I described above.

File: cpu/stm32l1/periph/i2c.c (hunk 12 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann What do you mean with automatically? In which case should it used for?

File: cpu/stm32l1/periph/i2c.c (hunk 6 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann Yes it's done [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32l1/periph/gpio.c#L80).

File: boards/limifrog-v1/include/periph_conf.h (hunk 2 above)
Comment (ReneHerthel) [inline boards/limifrog-v1/include/periph_conf.h]: So, keep it or remove it?

File: cpu/stm32l1/periph/i2c.c (hunk 5 above)
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: Hm? This is what the gaurd intents: `guard file in case no I2C device is defined`

File: cpu/stm32l1/periph/i2c.c (hunk 12 above)
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: If you won't use the statically defined macros here **but** enable the clock by an "intelligent" function here, you would save the whole switch case

File: cpu/stm32l1/periph/i2c.c (hunk 14 above)
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: @lebrush what do you mean byold behaviour (verification)? Do you mean `if (dev >= I2C_NUMOF)`?

**Comment by PeterKietzmann** [comment]
Sorry for commenting in that commit. Wasn't aware of it :/

File: cpu/stm32l1/periph/i2c.c (hunk 8 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann Okey, so there's a little problem. I was looking for the "right" errata sheet and found [this](http://www.st.com/web/en/resource/technical/document/errata_sheet/DM00104204.pdf). The problem is, when its copy-pasted from another code, it's maybe another errata sheet? I found nothing in chapter 2.14.7. Or I'm just blind :-P

There is something about I2C in chapter 2.4, but I'm not sure what we are searching for..

File: cpu/stm32l1/periph/i2c.c (hunk 14 above)
Comment (lebrush) [inline cpu/stm32l1/periph/i2c.c]: yes :-)

**Comment by ReneHerthel** [comment]
@PeterKietzmann Should I wait for the merge of the PR with the open drain you mentioned?

File: cpu/stm32l1/include/periph_cpu.h (hunk 4 above)
Comment (ReneHerthel) [inline cpu/stm32l1/include/periph_cpu.h]: Done.

File: cpu/stm32l1/periph/i2c.c (hunk 12 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann See my last commit

File: cpu/stm32l1/periph/i2c.c (hunk 14 above)
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: @ReneHerthel please put this check in again. We already discussed it some comments ago.

File: cpu/stm32l1/periph/i2c.c (hunk 8 above)
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: I mean: The toggling is **not** I2C specific. I suggested this is a copy-paste "error" from other MCUs. So if you don't find a hint in the corresponding errata sheet (for **your**) MCU it is possible that the toggling is not needed. Just do the test if everything goes as expected **without** it. In that case you can remove the whole shit

File: cpu/stm32l1/periph/i2c.c [hunk 15]
```diff
@@ -466,100 +362,105 @@ int i2c_write_regs(i2c_t dev, uint8_t address, uint8_t reg, char *data, int leng
...
-            I2C_0_CLKEN();
-            break;
-#endif
-#if I2C_1_EN
-        case I2C_1:
-            I2C_1_CLKEN();
-            break;
-#endif
-    }
+    RCC->APB1ENR |= (RCC_APB1ENR_I2C1EN << dev);
```
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: That was what I (generally) meant with removing the CLK enabling macros in the periph_conf - yes. But are you sure that this piece of code does what you expect? I'm not but I didn't check in detail

Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann Hmm [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32l1/periph/gpio.c#L72) is something similiar to this.

It uses the I2C-Enabling-Mask by left shifting it by the value of dev.

My thought:
`RCC_APB1ENR_I2C1EN`   := 0x00200000   := 0000 0000 0010 0000 .... and
`RCC_APB1ENR_I2C2EN` := 0x00400000 := 0000 0000 0100 0000 .... 
(see: [stm32l1xx.h](http://www.keil.com/dd/docs/arm/st/stm32l1xx/stm32l1xx.h))

for dev  = 0 it's the first one and for dev = 1 it's the second one by shifting RCC_APB1ENR_I2C1EN to left with 1.

If you have another idea how to do this better, please tell me :)

File: cpu/stm32l1/periph/i2c.c (hunk 14 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann It's already done

Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: Please include it in all functions.

File: cpu/stm32l1/periph/i2c.c (hunk 15 above)
Comment (PeterKietzmann) [inline cpu/stm32l1/periph/i2c.c]: Actually I was confused about `dev` but that was my bad. Seems reasonable.

Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann Check.

File: cpu/stm32l1/periph/i2c.c (hunk 14 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann Check.

File: cpu/stm32l1/periph/i2c.c (hunk 8 above)
Comment (ReneHerthel) [inline cpu/stm32l1/periph/i2c.c]: @PeterKietzmann Okey, seems to be working without it and I couldn't find something in the errata sheet.

**Comment by ReneHerthel** [comment]
Rebase.

**Comment by ReneHerthel** [comment]
Squashed

**Comment by PeterKietzmann** [comment]
@haukepetersen Travis fails IMO unrelated to this PR.
```
thread_msg:nrf51dongle:
Building application "thread_msg" for "nrf51dongle" with MCU "nrf51".
```
IIRC you did something with the nrf51 recently. Any ideas how to proceed?

**Comment by PeterKietzmann** [comment]
Forget about it, the build error seemed to be unrelated as we now have green lights. This change will need adaption to #4472 once a decision was found. We successfully tested both I2C ports with tests/driver_srf02 as well as with tests/periph_i2c (and the srf02 connected). Will merge **now** because we need this enhancement for further developments. ACK and go
